x-pack/filebeat/input/httpjson/request.go

```diff
@@ -285,27 +290,41 @@ func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, p
 		urlCopy                 url.URL
 		urlString               string
 		httpResp                *http.Response
-		initialResponse         []*http.Response
-		intermediateResps       []*http.Response
-		finalResps              []*http.Response
 		isChainWithPageExpected bool
 		chainIndex              int
 	)
+	initialResponseChannel := make(chan *http.Response, 1)
+	defer func(n *int) {
+		r.log.Infof("request finished: %d events published", *n)
+	}(&n)
 
 	for i, rf := range r.requestFactories {
-		finalResps = nil
-		intermediateResps = nil
+		publishedEventsCountChannel := make(chan int, 1)
+		events := make(chan maybeMsg, 1)
+		responseChannel := make(chan *http.Response, 1)
+		intermediateResponseChannel := make(chan *http.Response, 1)
+
 		// iterate over collected ids from last response
 		if i == 0 {
 			// perform and store regular call responses
 			httpResp, err = rf.collectResponse(stdCtx, trCtx, r)
 			if err != nil {
+				closeChannels(
+					[]chan int{publishedEventsCountChannel},
+					[]chan maybeMsg{events},
+					[]chan *http.Response{responseChannel, intermediateResponseChannel, initialResponseChannel},
+				)
```

Comment:
```suggestion
				close(publishedEventsCountChannel)
				close(events)
				closeChannels(responseChannel, intermediateResponseChannel, initialResponseChannel)
```
making use of the `closeChannels` below.